File: src/lib/converter/map-exporter/map-exporter.ts
import chalk from 'chalk';
import { existsSync, readdirSync, rmSync } from 'fs';
import fsExtra from 'fs-extra';
import _ from 'lodash';
import path from 'path';

import { exportCreatureModels } from '@/lib/azerothcore-client/creatures';
import { dataHeightMin, dataHeightToGameZ, maxGameHeightDiff } from '@/lib/constants';
import { WowObjectType } from '@/lib/converter/common/models';
import { WowObjectManager } from '@/lib/converter/common/wow-object-manager';
import { Wc3Converter } from '@/lib/converter/map-exporter/wc3-converter';
import { Config } from '@/lib/global-config';
import { Vector2 } from '@/lib/math/common';
import { radians } from '@/lib/math/rotation';
import { ModificationType } from '@/vendors/wc3maptranslator/data';
import { MapManager } from '@/vendors/wc3maptranslator/extra/map-manager';

export interface MapExportConfig {
  mapId: number;
  wowExportFolder: string;
  wmoSet?: string[];
  min: Vector2;
  max: Vector2;
  mapAngleDeg: number;

  terrain: {
    clampPercent: {
      upper: number,
      lower: number;
    },
  },

  doodads: {
    enable: Record<WowObjectType | 'others', boolean>;
    pitchRollThresholdRadians: number
  },

  creatures: {
    enable: boolean
    allAreDoodads: boolean;
    scaleUp: number
  },

}

export const defaultMapExportConfig: Omit<MapExportConfig, 'mapId' | 'wowExportFolder' | 'outputPath' | 'min' | 'max' | 'mapAngleDeg' > = {
  terrain: {
    clampPercent: {
      upper: 1,
      lower: 0,
    },
  },
  doodads: {
    enable: {
      adt: true,
      wmo: true,
      m2: true,
      gobj: true,
      others: true,
      unit: true,
    },
    pitchRollThresholdRadians: radians(5 /* degrees */),
  },
  creatures: {
    enable: true,
    allAreDoodads: false,
    scaleUp: 1,
  },
};

export class MapExporter {
  mapManager: MapManager;

  wowObjectManager: WowObjectManager;

  private filterDoodads: (id: string, type: WowObjectType) => boolean;

  constructor(public config: Config, public mapExportConfig: MapExportConfig) {
    this.mapManager = new MapManager();
  }

  public async parseObjects(filter?: (id: string, type: WowObjectType) => boolean) {
    const {
      wowExportFolder, min, max, mapAngleDeg, mapId,
    } = this.mapExportConfig;

    this.wowObjectManager = new WowObjectManager(this.config);
    this.filterDoodads = (id, type) => (this.mapExportConfig.doodads.enable[type] ?? this.mapExportConfig.doodads.enable.others) && type !== 'unit' && (filter?.(id, type) ?? true);
    if (this.mapExportConfig.wmoSet) {
      await this.wowObjectManager.readTerrainsDoodads(this.mapExportConfig.wmoSet, this.filterDoodads);
    }
    await this.wowObjectManager.readTerrainsDoodads(
      buildPaths(`**/${wowExportFolder}`, min, max),
      this.filterDoodads,
    );

    if (this.mapExportConfig.creatures.enable) {
      await this.wowObjectManager.readCreatures(mapId);
    }

    console.log('Total objects:', this.wowObjectManager.objects.size);
    const typeCountMap = _([...this.wowObjectManager.objects.values()])
      .map((o) => o.type)
      .countBy();
    console.log('Object type count:', typeCountMap.entries().toJSON());

    // Rotate all root objects
    console.log(`Rotating roots at center by ${mapAngleDeg} degrees`);
    this.wowObjectManager.rotateRootsAtCenter([0, 0, radians(-90 + mapAngleDeg)]);
  }

  public async exportTerrainsDoodads(outputDir: string) {
    const mapConfig = this.mapExportConfig;

    const wc3Converter = new Wc3Converter(mapConfig);
    this.mapManager.terrain = wc3Converter.generateTerrainWithHeight(this.wowObjectManager);

    // Place doodads
    const { doodadTypesWithPitchRoll } = wc3Converter.placeDoodads(
      this.mapManager,
      this.wowObjectManager,
      (obj) => this.filterDoodads(obj.id, obj.type),
    );

    console.log('Total doodads:', this.mapManager.doodads.length);
    console.log('Total doodad types:', this.mapManager.doodadTypes.length);
    console.log('Doodad types with custom pitch/roll:', doodadTypesWithPitchRoll);
    if (this.mapManager.doodads.length > 130_000) {
      throw new Error(`Too many doodads: ${this.mapManager.doodads.length}, limit is 130_000`);
    }

    // Export only assets that are actually used by placed doodads
    const am = this.wowObjectManager.assetManager;
    const usedModelPaths = new Set<string>();
    const collectModelPath = (mods: MapManager['doodadTypes'][number]['data']) => {
      // dfil for doodads, bfil for destructibles
      const file = mods.find((m) => (m.id === 'dfil' || m.id === 'bfil')
        && m.type === ModificationType.string
        && typeof m.value === 'string');
      if (file && typeof file.value === 'string') {
        usedModelPaths.add((file.value as string).replace(/\\/g, '/'));
      }
    };
    this.mapManager.doodadTypes.forEach((t) => collectModelPath(t.data));
    this.mapManager.destructibleTypes.forEach((t) => collectModelPath(t.data));

    // Keep only models that match the used model paths
    Array.from(am.models.entries()).forEach(([k, model]) => {
      const rel = model.mdl.model.name.replace(/\\/g, '/');
      if (!usedModelPaths.has(rel)) {
        am.models.delete(k);
      }
    });

    // Collect textures used by the remaining models and purge others
    const usedTexturePngPaths: string[] = [];
    am.models.forEach((model) => {
      model.mdl.textures.forEach((tex) => {
        // Derive from image (BLP) path and keep asset prefix if present
        const pngPath = tex.image ? tex.image.replace(/\.blp$/i, '.png').replace(/\\/g, '/') : '';
        if (pngPath) usedTexturePngPaths.push(pngPath);
      });
    });
    am.purgeTextures(usedTexturePngPaths);

    await am.exportTextures(outputDir);
    await am.exportModels(outputDir);
  }

  public async exportCreatures(outputDir: string) {
    const mapConfig = this.mapExportConfig;

    // Place creatures as Units or Doodads
    if (mapConfig.creatures.enable) {
      const wc3Converter = new Wc3Converter(mapConfig);
      const units = wc3Converter.placeUnits(this.mapManager, this.wowObjectManager);
      console.log('Created', this.mapManager.unitTypes.length, 'custom unit types');
      console.log('Placed', this.mapManager.units.length, 'unit instances');

      // Export creature assets
      const start = performance.now();
      units.sort((a, b) => a.creature.model.CreatureDisplayID - b.creature.model.CreatureDisplayID);
      await exportCreatureModels(units.map((u) => u.creature), outputDir, this.config);
      console.log('Exported all unit assets in', chalk.yellow(((performance.now() - start) / 1000).toFixed(2)), 's');
      console.log('Done');
    }
  }

  public saveWar3mapFiles(outputDir: string) {
    // Save map
    const templateEmptyDir = 'maps/template-empty.w3x';
    if (!existsSync(outputDir)) {
      fsExtra.copySync(templateEmptyDir, outputDir);
    } else {
      // copy all files in template-empty.w3x to outputDir if not already exists
      const allFiles = readdirSync(templateEmptyDir);
      for (const file of allFiles) {
        if (!existsSync(path.join(outputDir, file))) {
          fsExtra.copySync(path.join(templateEmptyDir, file), path.join(outputDir, file));
        }
      }
    }
    this.mapManager.save(outputDir);
    try {
      // Remove precomputed shadow file if it exists, since it no longers match objects on the map
      rmSync(path.join(outputDir, 'war3map.shd'));
    } catch (e) {
      // ignore
    }
  }
}

function buildPaths(prefix: string, min: Vector2, max: Vector2) {
  const res: string[] = [];
  for (let i = min[0]; i <= max[0]; i++) {
    for (let j = min[1]; j <= max[1]; j++) {
      res.push(path.join(prefix, `adt_${i}_${j}.obj`));
    }
  }
  return res;
}

export function gameZToPercent(z: number) {
  return (z - dataHeightToGameZ(dataHeightMin)) / maxGameHeightDiff;
}

File: golang/internal/converter/mapexporter/map_exporter.go
package mapexporter

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/pqhuy98/wow-converter/internal/ansi"
	"github.com/pqhuy98/wow-converter/internal/azerothcore"
	"github.com/pqhuy98/wow-converter/internal/config"
	"github.com/pqhuy98/wow-converter/internal/converter/character"
	"github.com/pqhuy98/wow-converter/internal/converter/common"
	"github.com/pqhuy98/wow-converter/internal/math"
	"github.com/pqhuy98/wow-converter/internal/wc3/data"
	"github.com/pqhuy98/wow-converter/internal/wc3/extra"
	"github.com/pqhuy98/wow-converter/internal/workspace"
	"github.com/pqhuy98/wow-converter/internal/wow/client"
)

// MapExportConcurrency caps parallel workers for map export tasks (ADT tiles, creatures).
const MapExportConcurrency = 8

// MapExportWorkerCount returns bounded worker count for map export tasks.
func MapExportWorkerCount() int {
	n := config.MaxConcurrency()
	if n > MapExportConcurrency {
		return MapExportConcurrency
	}
	return n
}

// TerrainClampPercent configures Z clamp for terrain generation.
type TerrainClampPercent struct {
	Lower float64
	Upper float64
}

// DoodadEnable toggles doodad types during export.
type DoodadEnable struct {
	ADT    bool
	WMO    bool
	M2     bool
	Gobj   bool
	Unit   bool
	Others bool
}

// MapExportConfig drives a single map export run.
type MapExportConfig struct {
	MapID           int
	WowExportFolder string
	WMOSet          []string
	Min             math.Vector2
	Max             math.Vector2
	MapAngleDeg     float64
	Terrain         struct {
		ClampPercent TerrainClampPercent
	}
	Doodads struct {
		Enable                    DoodadEnable
		PitchRollThresholdRadians float64
	}
	Creatures struct {
		Enable        bool
		AllAreDoodads bool
	}
	UnitScale float64
}

// DefaultMapExportConfig returns defaults excluding map-specific fields.
func DefaultMapExportConfig() MapExportConfig {
	cfg := MapExportConfig{
		Terrain: struct{ ClampPercent TerrainClampPercent }{
			ClampPercent: TerrainClampPercent{Lower: 0, Upper: 1},
		},
		UnitScale: 1,
	}
	cfg.Doodads.Enable = DoodadEnable{ADT: true, WMO: true, M2: true, Gobj: true, Unit: true, Others: true}
	cfg.Doodads.PitchRollThresholdRadians = math.Radians(5)
	cfg.Creatures.Enable = true
	return cfg
}

// MapExporter orchestrates WoW -> WC3 map conversion.
type MapExporter struct {
	Config           config.Config
	MapExportConfig  *MapExportConfig
	MapManager       *extra.MapManager
	WowObjectManager *common.WowObjectManager
	wowClient        client.Client
	tileRegistry     *common.TileRegistry
	filterDoodads    func(id string, typ common.WowObjectType) bool
}

// NewMapExporter creates an exporter. mapCfg must be non-nil and is mutated in place during export.
func NewMapExporter(cfg config.Config, mapCfg *MapExportConfig, wowClient client.Client, tileRegistry *common.TileRegistry) *MapExporter {
	if mapCfg == nil {
		panic("mapexporter: nil MapExportConfig")
	}
	return &MapExporter{
		Config:          cfg,
		MapExportConfig: mapCfg,
		MapManager:      extra.NewMapManager(),
		wowClient:       wowClient,
		tileRegistry:    tileRegistry,
	}
}

// ParseObjects reads terrains, doodads, and creatures.
func (e *MapExporter) ParseObjects(filter func(id string, typ common.WowObjectType) bool) error {
	mc := e.MapExportConfig
	e.WowObjectManager = common.NewWowObjectManager(e.Config, e.wowClient, e.tileRegistry)
	e.filterDoodads = func(id string, typ common.WowObjectType) bool {
		enabled := mc.Doodads.Enable.Others
		switch typ {
		case common.WowObjectADT:
			enabled = mc.Doodads.Enable.ADT
		case common.WowObjectWMO:
			enabled = mc.Doodads.Enable.WMO
		case common.WowObjectM2:
			enabled = mc.Doodads.Enable.M2
		case common.WowObjectGobj:
			enabled = mc.Doodads.Enable.Gobj
		case common.WowObjectUnit:
			enabled = mc.Doodads.Enable.Unit
		}
		return enabled && typ != common.WowObjectUnit && (filter == nil || filter(id, typ))
	}

	if len(mc.WMOSet) > 0 {
		if err := e.WowObjectManager.ReadTerrainsDoodads(mc.WMOSet, e.filterDoodads); err != nil {
			return err
		}
	}
	paths := buildPaths("**/"+mc.WowExportFolder, mc.Min, mc.Max)
	if err := e.WowObjectManager.ReadTerrainsDoodads(paths, e.filterDoodads); err != nil {
		return err
	}
	if err := e.WowObjectManager.ReadCreatures(mc.MapID); err != nil {
		return err
	}

	log.Printf("Total objects: %d", len(e.WowObjectManager.Objects))
	typeCount := map[string]int{}
	for _, obj := range e.WowObjectManager.Objects {
		typeCount[string(obj.Type)]++
	}
	log.Printf("Object type count: %v", typeCount)
	log.Printf("Rotating roots at center by %.0f degrees", mc.MapAngleDeg)
	e.WowObjectManager.RotateRootsAtCenter([3]float64{0, 0, math.Radians(-90 + mc.MapAngleDeg)})
	if e.tileRegistry != nil {
		e.tileRegistry.TrimAfterParse()
	}
	return nil
}

// ExportTerrainsDoodads generates terrain and exports doodad assets.
func (e *MapExporter) ExportTerrainsDoodads(outputDir string) error {
	wc3 := NewWc3Converter(*e.MapExportConfig)
	e.MapManager.SetTerrain(wc3.GenerateTerrainWithHeight(e.WowObjectManager))

	_, _ = wc3.PlaceDoodads(e.MapManager, e.WowObjectManager, func(obj *common.WowObject) bool {
		return e.filterDoodads(obj.ID, obj.Type)
	})
	log.Printf("Total doodads: %d", len(e.MapManager.Doodads))
	log.Printf("Total doodad types: %d", len(e.MapManager.DoodadTypes))
	pitchRollCount := 0
	for _, t := range e.MapManager.DoodadTypes {
		for _, m := range t.Data {
			if m.ID == "dprx" || m.ID == "dpro" {
				pitchRollCount++
				break
			}
		}
	}
	log.Printf("Doodad types with custom pitch/roll: %d", pitchRollCount)

	if len(e.MapManager.Doodads) > 130000 {
		return fmt.Errorf("too many doodads: %d, limit is 130000", len(e.MapManager.Doodads))
	}

	am := e.WowObjectManager.AssetManager
	usedModelPaths := map[string]struct{}{}
	collectModelPath := func(mods []data.Modification) {
		for _, m := range mods {
			if (m.ID == "dfil" || m.ID == "bfil") && m.Type == data.ModificationString {
				if s, ok := m.Value.(string); ok {
					usedModelPaths[strings.ReplaceAll(s, "\\", "/")] = struct{}{}
				}
			}
		}
	}
	for _, t := range e.MapManager.DoodadTypes {
		collectModelPath(t.Data)
	}
	for _, t := range e.MapManager.DestructibleTypes {
		collectModelPath(t.Data)
	}
	for k, model := range am.Models() {
		rel := strings.ReplaceAll(model.MDL.Model.Name, "\\", "/")
		if _, ok := usedModelPaths[rel]; !ok {
			delete(am.Models(), k)
		}
	}

	var usedTextures []string
	for _, model := range am.Models() {
		for _, tex := range model.MDL.Textures {
			if tex.WowData.PngPath != "" {
				usedTextures = append(usedTextures, strings.ReplaceAll(tex.WowData.PngPath, "\\", "/"))
			} else if tex.Image != "" {
				usedTextures = append(usedTextures, strings.ReplaceAll(strings.ReplaceAll(tex.Image, ".blp", ".png"), "\\", "/"))
			}
		}
	}
	am.PurgeTextures(usedTextures)

	if _, err := am.ExportTextures(outputDir); err != nil {
		return err
	}
	if err := am.ExportModels(outputDir); err != nil {
		return err
	}
	am.ReleaseAfterExport()
	return nil
}

// ExportCreatures places units and exports creature models.
func (e *MapExporter) ExportCreatures(outputDir string, onProgress func(completed, total int)) error {
	mc := e.MapExportConfig
	if !mc.Creatures.Enable {
		return nil
	}
	wc3 := NewWc3Converter(*mc)
	units := wc3.PlaceUnits(e.MapManager, e.WowObjectManager)
	log.Printf("Created %d custom unit types", len(e.MapManager.UnitTypes))
	log.Printf("Placed %d unit instances", len(e.MapManager.Units))

	creatures := make([]azerothcore.Creature, 0, len(units))
	for _, u := range units {
		creatures = append(creatures, u.Creature)
	}
	start := time.Now()
	if err := character.ExportCreatureModels(creatures, outputDir, e.Config, e.wowClient, MapExportWorkerCount(), onProgress); err != nil {
		return err
	}
	log.Printf("Exported all unit assets in %s", ansi.Yellowf("%.2fs", time.Since(start).Seconds()))
	log.Printf("Done")
	return nil
}

// SaveWar3mapFiles copies template and writes map binaries.
func (e *MapExporter) SaveWar3mapFiles(outputDir, mapName string) error {
	templateDir, err := workspace.ResolveTemplateEmptyDir()
	if err != nil {
		return err
	}
	if _, err := os.Stat(outputDir); os.IsNotExist(err) {
		if err := copyDir(templateDir, outputDir); err != nil {
			return err
		}
	} else {
		entries, _ := os.ReadDir(templateDir)
		for _, ent := range entries {
			dst := filepath.Join(outputDir, ent.Name())
			if _, err := os.Stat(dst); os.IsNotExist(err) {
				if err := copyFile(filepath.Join(templateDir, ent.Name()), dst); err != nil {
					return err
				}
			}
		}
	}
	e.MapManager.EnsureMapInfo(mapName)
	if err := e.MapManager.Save(outputDir); err != nil {
		return err
	}
	_ = os.Remove(filepath.Join(outputDir, "war3map.shd"))
	return nil
}

func buildPaths(prefix string, min, max math.Vector2) []string {
	var res []string
	for i := int(min[0]); i <= int(max[0]); i++ {
		for j := int(min[1]); j <= int(max[1]); j++ {
			res = append(res, filepath.Join(prefix, fmt.Sprintf("adt_%d_%d.obj", i, j)))
		}
	}
	return res
}

func copyDir(src, dst string) error {
	return filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		rel, _ := filepath.Rel(src, path)
		target := filepath.Join(dst, rel)
		if info.IsDir() {
			return os.MkdirAll(target, 0o755)
		}
		return copyFile(path, target)
	})
}

func copyFile(src, dst string) error {
	data, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return err
	}
	return os.WriteFile(dst, data, 0o644)
}
